File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    행 = len(maps)
    열 = len(maps[0])
    dx = [0, 0, -1, +1]  # 상하좌우
    dy = [+1, -1, 0, 0]
    
    q = deque()
    visited = [[False] * 열 for _ in range(행)]
    
    q.append((0, 0))
    visited[0][0] = True
    
    while q:
        x, y = q.popleft()
        for m, n in zip(dx, dy):
            nx = x + m
            ny = y + n
            if 0 <= nx < 행 and 0 <= ny < 열:
                if maps[nx][ny] == 1 and not visited[nx][ny]:
                    q.append((nx, ny))
                    visited[nx][ny] = True
                    maps[nx][ny] = maps[x][y] + 1
                    
    return maps[행 - 1][열 - 1] if maps[행 - 1][열 - 1] != 1 else -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<bits/stdc++.h>

using namespace std;

int solution(vector<vector<int>> maps)
{
    int m = maps.size(); //세로
    int n = maps[0].size(); //가로
    
    queue<pair<int,int>> q; //BFS이니까 먼저 도착한 게 최단경로
    vector<vector<bool>> visited(m,vector<bool>(n, false));
    
    visited[0][0] = true;
    q.push({0,0});
    
    vector<int> dy = {-1, 1, 0, 0};
    vector<int> dx = {0, 0, -1, 1};
    
    // 가능한 길을 계속 나가기
    while(q.size() > 0){
       auto [y, x] = q.front();
        q.pop();
        
        //종료조건
        if ((y == m-1) && (x == n-1)){
            return maps[y][x];
        }
        
        for(int i = 0; i < 4; i++){
            int ny = y + dy[i];
            int nx = x + dx[i];
            
            if((0 <= ny && ny < m  && 0 <= nx && nx < n) && (maps[ny][nx] == 1) && (!visited[ny][nx])){
                maps[ny][nx] = maps[y][x] + 1;
                visited[ny][nx] = true;
                q.push({ny,nx});
            }
        }
    }
    
    // 만약 q.size() == 0인데 도착하지 못했으면 
    return -1;
}